File: cpp/image-acquisition/record-mp4-h264/src/record-mp4-h264.cpp

#include <ic4/ic4.h>

#include <console-helper.h>

#include <iostream>
#include <cstdio>
#include <atomic>

// Define QueueSinkListener-derived class that can add the received frames into a video writer
class AddFrameListener : public ic4::QueueSinkListener
{
private:
	ic4::VideoWriter& writer_;
	std::atomic<bool> do_write_frames_;
	std::atomic<int> num_frames_written_;

public:
	AddFrameListener(ic4::VideoWriter& writer)
		: writer_(writer)
		, do_write_frames_(false)
		, num_frames_written_(0)
	{
	}

	// Inherited via QueueSinkListener, called when there are frames available in the sink's output queue
	void framesQueued(ic4::QueueSink& sink) override
	{
		ic4::Error err;

		// Remove a buffer from the sink's output queue
		// We have to remove buffers from the queue even if not recording; otherwise the device will not have
		// buffers to write new video data into
		auto buffer = sink.popOutputBuffer(err);
		if (buffer == nullptr)
		{
			std::cerr << "Failed to get frame from sink: " << err.message() << std::endl;
			return;
		}

		if (do_write_frames_)
		{
			// Pass the image buffer to the video writer
			if (!writer_.addFrame(buffer, err))
			{
				std::cerr << "Failed to add frame to video file: " << err.message() << std::endl;
			}

			num_frames_written_ += 1;
		}
	}

	void enable_recording(bool enable)
	{
		if (enable)
		{
			num_frames_written_ = 0;
		}
		do_write_frames_ = enable;
	}

	int num_frames_written()
	{
		return num_frames_written_;
	}
};

int main()
{
	ic4::initLibrary();
	std::atexit(ic4::exitLibrary);

	// Let the user select a device
	auto device_list = ic4::DeviceEnum::enumDevices();
	auto it = ic4_examples::console::select_from_list(device_list);
	if (it == device_list.end())
	{
		return -1;
	}

	// Open the selected device
	ic4::Error err;
	ic4::Grabber grabber;
	if (!grabber.deviceOpen(*it, err))
	{
		std::cerr << "Failed to open device: " << err.message() << std::endl;
		return -2;
	}

	// #TODO: Insert format configuration

	// Create a video writer for H264-compressed MP4 files
	ic4::VideoWriter writer(ic4::VideoWriterType::MP4_H264);

	// Create an instance of the listener type defined above
	AddFrameListener listener(writer);

	// Create a QueueSink to capture all images arriving from the video capture device
	auto sink = ic4::QueueSink::create(listener, err);
	if (!sink)
	{
		std::cerr << "Failed to create sink: " << err.message() << std::endl;
		return -3;
	}

	// Start the video stream into the sink
	if (!grabber.streamSetup(sink, ic4::StreamSetupOption::AcquisitionStart, err))
	{
		std::cerr << "Failed to setup stream: " << err.message() << std::endl;
		return -4;
	}

	// Query the sink's output image type
	// The image type is required when starting the video recording
	auto image_type = sink->outputImageType(err);
	if (err.isError())
	{
		std::cerr << "Failed to query sink output frame type: " << err.message() << std::endl;
		return -5;
	}

	// Query the device's configured frame rate.
	// The frame rate is later set as the video file's playback rate.
	auto frame_rate = grabber.devicePropertyMap().find(ic4::PropId::AcquisitionFrameRate).getValue(err);
	if (err.isError())
	{
		std::cerr << "Failed to query acquisition frame rate: " << err.message() << std::endl;
		return -6;
	}

	std::cout << "Stream started." << std::endl;
	std::cout << "ImageType: " << ic4::to_string(image_type) << std::endl;
	std::cout << "AcquisitionFrameRate: " << frame_rate << std::endl;
	std::cout << std::endl;

	for (int i = 0; i < 3; ++i)
	{
		std::cout << "Press any key to begin recording a video file" << std::endl;
		(void)std::getchar();

		std::string file_name = "video" + std::to_string(i) + ".mp4";

		// Begin writing a video file with a name, image type and playback rate
		if (!writer.beginFile(file_name.c_str(), image_type, frame_rate, err))
		{
			std::cerr << "Failed to begin recording: " << err.message() << std::endl;
			continue;
		}

		// Instruct our QueueSinkListener to write frames into the video writer
		listener.enable_recording(true);

		std::cout << "Recording started. Press any key to stop" << std::endl;
		(void)std::getchar();

		// Stop writing frames into the video writer
		listener.enable_recording(false);

		// Finalize the currently opened video file
		if (!writer.finishFile(err))
		{
			std::cerr << "Failed to finish recording: " << err.message() << std::endl;
			continue;
		}

		std::cout << "Saved video file " << file_name << std::endl;
		std::cout << "Wrote " << listener.num_frames_written() << " frames." << std::endl;
		std::cout << std::endl;
	}

	// We have to call streamStop before exiting the function, because we have the listener defined as a stack variable.
	grabber.streamStop();
	grabber.deviceClose();

	return 0;
}

File: python/image-acquisition/record-mp4-h264/record-mp4-h264.py

import imagingcontrol4 as ic4


def example_record_mp4_h264():
    # Let the user select one of the connected cameras
    device_list = ic4.DeviceEnum.devices()
    for i, dev in enumerate(device_list):
        print(f"[{i}] {dev.model_name} ({dev.serial}) [{dev.interface.display_name}]")
    print(f"Select device [0..{len(device_list) - 1}]: ", end="")
    selected_index = int(input())
    dev_info = device_list[selected_index]

    # Open the selected device in a new Grabber
    grabber = ic4.Grabber(dev_info)
    map = grabber.device_property_map

    # Reset all device settings to default
    # Not all devices support this, so ignore possible errors
    map.try_set_value(ic4.PropId.USER_SET_SELECTOR, "Default")
    map.try_set_value(ic4.PropId.USER_SET_LOAD, 1)

    # Create a video writer for H264-compressed MP4 files
    video_writer = ic4.VideoWriter(ic4.VideoWriterType.MP4_H264)

    # Define a listener class to receive queue sink notifications
    # The listener will pass incoming frames to the video writer
    class Listener(ic4.QueueSinkListener):
        def __init__(self, video_writer: ic4.VideoWriter):
            self.video_writer = video_writer
            self.counter = 0
            self.do_write_frames = False

        def sink_connected(self, sink: ic4.QueueSink, image_type: ic4.ImageType, min_buffers_required: int) -> bool:
            # No need to configure anything, just accept the connection
            return True

        def frames_queued(self, sink: ic4.QueueSink):
            # Get the queued image buffer
		    # We have to remove buffers from the queue even if not recording; otherwise the device will not have
		    # buffers to write new video data into
            buffer = sink.pop_output_buffer()

            if self.do_write_frames:
                self.video_writer.add_frame(buffer)
                self.counter = self.counter + 1

        def enable_recording(self, enable: bool):
            if enable:
                self.counter = 0
            self.do_write_frames = enable

        def num_frames_written(self):
            return self.counter

    # Create an instance of the listener type defined above, specifying a partial file name
    listener = Listener(video_writer)

    # Create a QueueSink to capture all images arriving from the video capture device
    sink = ic4.QueueSink(listener)

    # Start the video stream into the sink
    grabber.stream_setup(sink)

    image_type = sink.output_image_type
    frame_rate = map[ic4.PropId.ACQUISITION_FRAME_RATE].value

    print("Stream started.")
    print(f"ImageType: {image_type}")
    print(f"AcquisitionFrameRate {frame_rate}")
    print()

    for i in range(3):
        input("Press ENTER to begin recording a video file")

        file_name = f"video{i}.mp4"

        # Begin writing a video file with a name, image type and playback rate
        video_writer.begin_file(file_name, image_type, frame_rate)

        # Instruct our QueueSinkListener to write frames into the video writer
        listener.enable_recording(True)

        input("Recording started. Press ENTER to stop")

        # Stop writing frames into the video writer
        listener.enable_recording(False)

        # Finalize the currently opened video file
        video_writer.finish_file()

        print(f"Saved video file {file_name}.")
        print(f"Wrote {listener.num_frames_written()} frames.")
        print()


    # We have to call streamStop before exiting the function, to make sure the listener object is not destroyed before the stream is stopped
    grabber.stream_stop()
    grabber.device_close()

if __name__ == "__main__":
    with ic4.Library.init_context(api_log_level=ic4.LogLevel.INFO, log_targets=ic4.LogTarget.STDERR):

        example_record_mp4_h264()